File: typescript/src/02.ts
import { readFileSync } from 'fs';

const input = readFileSync ('../data/02.txt', 'utf-8')
    .trimEnd(); // remove trailing newline

const limits: { [key: string]: number } = { red: 12, green: 13, blue: 14 };

let total = 0
input.split('\n').forEach(line => {
    const match = line.match(/Game (\d+): (.*)/);
    if (!match) { throw new Error(); }
    const num = parseInt(match[1]);
    const batches = match[2];
    let possible = true;

    batches.split('; ').forEach(batch => {
        batch.split(', ').forEach(item => {
            let itemMatch = item.match(/(\d+) (\w+)/)
            if (!itemMatch) { throw new Error(); }
            const count = parseInt(itemMatch[1]);
            const color = itemMatch[2];
            if (count > limits[color]) {
                possible = false;
            }
        });
    });

    if (possible) {
        total += num;
    }
});
console.log (total)

File: lua/src/02.lua
local input = io.open('../data/02.txt'):read('*a')

local limits = { red = 12, green = 13, blue = 14 }

local total = 0

for line in input:gmatch ('[^\n]+') do
  local num, batches = line:match 'Game (%d+): (.*)'
  num = tonumber(num)
  local possible = true
  for batch in batches:gmatch '[^;]+' do
    for item in batch:gmatch '[^,]+' do
      local count, color = item:match '(%d+) (%w+)'
      if tonumber(count) > limits [color] then
        possible = false
      end
    end
  end
  if possible then
    total = total + num
  end
end

print (total)
